File: app/src/main/java/com/lodz/android/hermesdemo/WebSocketActivity.java
package com.lodz.android.hermesdemo;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.core.widget.NestedScrollView;

import com.lodz.android.component.base.activity.BaseActivity;
import com.lodz.android.component.widget.base.TitleBarLayout;
import com.lodz.android.core.utils.DateUtils;
import com.lodz.android.core.utils.ScreenUtils;
import com.lodz.android.core.utils.StatusBarUtil;
import com.lodz.android.core.utils.ToastUtils;
import com.lodz.android.hermes.contract.Hermes;
import com.lodz.android.hermes.contract.OnConnectListener;
import com.lodz.android.hermes.contract.OnSendListener;
import com.lodz.android.hermes.contract.OnSubscribeListener;
import com.lodz.android.hermes.modules.HermesAgent;

import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * WebSocket测试类
 * @author zhouL
 * @date 2019/5/6
 */
public class WebSocketActivity extends BaseActivity {

    public static void start(Context context) {
        Intent starter = new Intent(context, WebSocketActivity.class);
        context.startActivity(starter);
    }

    /** 默认地址 */
    private static final String DEFAULT_URL = "ws://192.168.7.39:9090/yszz/jms/1";

    /** 地址输入框 */
    @BindView(R.id.url_edit)
    EditText mUrlEdit;
    /** 连接按钮 */
    @BindView(R.id.connect_btn)
    Button mConnectBtn;
    /** 发送内容输入框 */
    @BindView(R.id.send_edit)
    EditText mSendEdit;
    /** 发送按钮 */
    @BindView(R.id.send_btn)
    Button mSendBtn;

    /** 滚动控件 */
    @BindView(R.id.scroll_view)
    NestedScrollView mScrollView;
    /** 日志 */
    @BindView(R.id.result_tv)
    TextView mResultTv;

    /** 清空按钮 */
    @BindView(R.id.clean_btn)
    Button mCleanBtn;
    /** 断开按钮 */
    @BindView(R.id.disconnect_btn)
    Button mDisconnectBtn;

    /** 日志 */
    private String mLog = "";
    /** 推送客户端 */
    private Hermes mHermes;
    /** 保活订阅者 */
    private Disposable mDisposable;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_ws;
    }

    @Override
    protected void findViews(Bundle savedInstanceState) {
        ButterKnife.bind(this);
        initTitleBarLayout(getTitleBarLayout());
        StatusBarUtil.setColor(this, ContextCompat.getColor(getContext(), R.color.colorPrimary), 0);
    }

    private void initTitleBarLayout(TitleBarLayout titleBarLayout) {
        titleBarLayout.setTitleName(R.string.ws_title);
        titleBarLayout.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.colorPrimary));
        titleBarLayout.setTitleTextColor(R.color.white);
        titleBarLayout.needBackButton(false);
    }

    @Override
    protected boolean onPressBack() {
        if (mHermes != null){
            if (mHermes.isConnected()){
                mHermes.disconnect();
            }
            mHermes = null;
            mDisposable.dispose();
        }
        return super.onPressBack();
    }

    @Override
    protected void setListeners() {
        super.setListeners();

        // 连接按钮
        mConnectBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String url = mUrlEdit.getText().toString();
                if (TextUtils.isEmpty(url)) {
                    ToastUtils.showShort(getContext(), R.string.mqtt_url_hint);
                    return;
                }
                if (mHermes != null){
                    if (!mHermes.isConnected()){
                        mHermes.connect();
                    }
                    return;
                }
                mHermes = HermesAgent.create()
                        .setConnectType(HermesAgent.WEB_SOCKET)
                        .setUrl(url)
                        .setPrintLog(true)
                        .setOnConnectListener(new OnConnectListener() {
                            @Override
                            public void onConnectComplete(boolean isReconnected) {
                                logResult("连接完成 ： isReconnected ---> " + isReconnected);
                            }

                            @Override
                            public void onConnectFailure(Throwable cause) {
                                logResult("连接失败 : " + cause.getMessage());
                            }

                            @Override
                            public void onConnectionLost(Throwable cause) {
                                logResult("连接断开（丢失） : " + cause.getMessage());
                            }
                        })
                        .setOnSendListener(new OnSendListener() {
                            @Override
                            public void onSendComplete(String topic, String content) {
                                logResult("发送成功 : topic ---> " + topic + "   " + content);
                            }

                            @Override
                            public void onSendFailure(String topic, Throwable cause) {
                                logResult("发送失败 : topic ---> " + topic + "    " + cause.getMessage());
                            }
                        })
                        .setOnSubscribeListener(new OnSubscribeListener() {

                            @Override
                            public void onSubscribeSuccess(String topic) {
                                logResult("订阅成功 : topic ---> " + topic);
                            }

                            @Override
                            public void onSubscribeFailure(String topic, Throwable cause) {
                                logResult("订阅失败 : topic ---> " + topic + "    " + cause.getMessage());
                            }

                            @Override
                            public void onMsgArrived(String subTopic, String msg) {
                                logResult("消息到达(" + subTopic + ")： " + msg);
                            }
                        })
                        .buildConnect(getContext().getApplicationContext());
                initKeepAlive();
            }
        });

        // 发送按钮
        mSendBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String content = mSendEdit.getText().toString();
                if (TextUtils.isEmpty(content)) {
                    ToastUtils.showShort(getContext(), R.string.mqtt_send_content_empty);
                    return;
                }
                if (mHermes == null || !mHermes.isConnected()) {
                    ToastUtils.showShort(getContext(), R.string.mqtt_client_unconnected);
                    return;
                }
                mHermes.sendTopic("", mSendEdit.getText().toString());
            }
        });

        // 清空按钮
        mCleanBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mLog = "";
                mResultTv.setText("");
            }
        });

        // 断开按钮
        mDisconnectBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mHermes != null){
                    mHermes.disconnect();
                }
            }
        });
    }

    /** 初始化保活线程 */
    private void initKeepAlive() {
        mDisposable = Observable.interval(60, 60, TimeUnit.SECONDS)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(new Consumer<Long>() {
                    @Override
                    public void accept(Long i) throws Exception {
                        if (mHermes == null) {
                            if (mDisposable != null) {
                                mDisposable.dispose();
                            }
                            return;
                        }
                        if (!mHermes.isConnected()) {
                            mHermes.disconnect();
                            Thread.sleep(1000);
                            mHermes.connect();
                        }
                    }
                });
    }

    @Override
    protected void initData() {
        super.initData();
        mUrlEdit.setText(DEFAULT_URL);
        showStatusCompleted();
    }

    /**
     * 打印信息
     * @param result 信息
     */
    private void logResult(String result){
        if (TextUtils.isEmpty(result)) {
            result = "";
        }
        mLog = mLog + DateUtils.getCurrentFormatString(DateUtils.TYPE_8) + " : " + result + "\n";
        mResultTv.setText(mLog);
        mScrollView.post(new Runnable() {
            @Override
            public void run() {
                mScrollView.smoothScrollTo(ScreenUtils.getScreenWidth(getContext()), ScreenUtils.getScreenHeight(getContext()));
            }
        });
    }

}

File: app/src/main/java/com/lodz/android/hermesdemo/WebSocketActivity.kt
package com.lodz.android.hermesdemo

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import com.lodz.android.corekt.anko.*
import com.lodz.android.corekt.utils.DateUtils
import com.lodz.android.corekt.utils.StatusBarUtil
import com.lodz.android.hermes.contract.HermesWebSocketClient
import com.lodz.android.hermes.modules.HermesAgent
import com.lodz.android.hermes.ws.client.OnConnectListener
import com.lodz.android.hermes.ws.client.OnSendListener
import com.lodz.android.hermes.ws.client.OnSubscribeListener
import com.lodz.android.hermesdemo.databinding.ActivityWsBinding
import com.lodz.android.pandora.base.activity.BaseActivity
import com.lodz.android.pandora.utils.viewbinding.bindingLayout
import com.lodz.android.pandora.widget.base.TitleBarLayout
import java.nio.ByteBuffer

/**
 * WebSocket测试类
 * @author zhouL
 * @date 2020/3/17
 */
class WebSocketActivity : BaseActivity() {

    companion object {

        private const val DEFAULT_URL = "ws://124.222.224.186:8800"

        fun start(context: Context){
            context.startActivity(Intent(context, WebSocketActivity::class.java))
        }
    }

    private val mBinding: ActivityWsBinding by bindingLayout(ActivityWsBinding::inflate)

    override fun getViewBindingLayout(): View = mBinding.root

    /** WebSocket客户端 */
    private var mHermes: HermesWebSocketClient? = null

    override fun findViews(savedInstanceState: Bundle?) {
        super.findViews(savedInstanceState)
        initTitleBarLayout(getTitleBarLayout())
        StatusBarUtil.setColor(window, getColorCompat(R.color.colorPrimary))
    }

    private fun initTitleBarLayout(titleBarLayout: TitleBarLayout) {
        titleBarLayout.setTitleName(R.string.ws_title)
        titleBarLayout.setBackgroundColor(getColorCompat(R.color.colorPrimary))
        titleBarLayout.setTitleTextColor(R.color.white)
    }

    override fun onClickBackBtn() {
        releaseHermes()
        finish()
    }

    override fun onPressBack(): Boolean {
        releaseHermes()
        return false
    }

    private fun releaseHermes(){
        mHermes?.release()
        mHermes = null
    }

    override fun setListeners() {
        super.setListeners()
        // 连接按钮
        mBinding.createBtn.setOnClickListener {
            val url = mBinding.urlEdit.text.toString()
            if (url.isEmpty()) {
                toastShort(R.string.mqtt_url_hint)
                return@setOnClickListener
            }
            if (mHermes != null) {
                releaseHermes()
            }
            mHermes = HermesAgent.createWebSocketClient()
                .init(getContext())
                .setLogTag("WebSocketLog")
                .setPrintLog(true)
                .setSilent(false)
                .setReconnectInterval(20 * 1000)
                .setOnConnectListener(object : OnConnectListener {
                    override fun onConnectComplete() {
                        logResult("连接完成")
                    }

                    override fun onConnectFailure(cause: Throwable) {
                        logResult("连接失败 : ${cause.message}")
                    }

                    override fun onConnectionLost(cause: Throwable) {
                        logResult("连接断开（丢失） : ${cause.message}")
                    }
                })
                .setOnSubscribeListener(object : OnSubscribeListener {
                    override fun onMsgArrived(ip: String, msg: String) {
                        logResult("消息到达 : $ip ---> $msg")
                    }

                    override fun onMsgArrived(ip: String, msg: ByteBuffer) {
                        logResult("消息到达 : $ip ---> $msg")
                    }
                })
                .setOnSendListener(object : OnSendListener {
                    override fun onSendComplete(ip: String, msg: String) {
                        logResult("消息发送成功 : $ip -> $msg")
                    }

                    override fun onSendComplete(ip: String, msg: ByteBuffer) {
                        logResult("消息发送成功 : $ip -> $msg")
                    }

                    override fun onSendFailure(cause: Throwable) {
                        logResult("消息发送失败：${cause.message}")
                    }
                })
                .build(url, true)
            mHermes?.connect()
            logResult("开始连接：$url")
        }

        // 发送按钮
        mBinding.sendBtn.setOnClickListener {
            val content = mBinding.sendEdit.text.toString()
            if (content.isEmpty()) {
                toastShort(R.string.mqtt_send_content_empty)
                return@setOnClickListener
            }
            if (mHermes == null || mHermes?.isConnected() == false) {
                toastShort(R.string.mqtt_client_unconnected)
                return@setOnClickListener
            }
            mHermes?.sendData(mBinding.sendEdit.text.toString())
            mBinding.sendEdit.hideInputMethod()
        }

        // 清空日志按钮
        mBinding.cleanBtn.setOnClickListener {
            mBinding.resultTv.text = ""
        }

        // 连接按钮
        mBinding.connectBtn.setOnClickListener {
            val url = mBinding.urlEdit.text.toString()
            if (url.isEmpty()) {
                toastShort(R.string.mqtt_url_hint)
                return@setOnClickListener
            }
            if (mHermes?.isConnected().check()){
                logResult("已经连接")
                return@setOnClickListener
            }
            logResult("开始连接 : $url")
            mHermes?.build(url, true)
            mHermes?.connect()
        }

        // 断开按钮
        mBinding.disconnectBtn.setOnClickListener {
            mHermes?.disconnect()
        }

        mBinding.silentBtn.setOnClickListener {
            mHermes?.setSilent(true)
            logResult("静默，关闭消息到达提醒")
        }

        mBinding.unsilentBtn.setOnClickListener {
            mHermes?.setSilent(false)
            logResult("非静默，打开消息到达提醒")
        }
    }

    override fun initData() {
        super.initData()
        mBinding.urlEdit.setText(DEFAULT_URL)
        showStatusCompleted()
    }

    /** 打印信息[result] */
    private fun logResult(result: String) {
        val log = "【${DateUtils.getCurrentFormatString(DateUtils.TYPE_8)}】 ".append(result)
        val text = mBinding.resultTv.text
        if (text.isEmpty()) {
            mBinding.resultTv.text = log
        }
        mBinding.resultTv.text = log.append("\n").append(text)
    }
}